### PR TITLE
commitWithPermit

### PR DESCRIPTION
**Pull request by MrChico**
Allows for 1 tx registration

- inline mkCommitment
- add commitWithPermit
- fover RadicleToken over ERC20Burnable interface

### CODE REVIEW

File: contracts/Registrar.sol [hunk 1]
```diff
@@ -109,31 +102,44 @@ contract Registrar {
...
+    /// Commit to a future name and submit permit in the same transaction
+    function commitWithPermit(bytes32 commitment, address owner, address spender, uint256 value,
+                              uint256 deadline, uint8 v, bytes32 r, bytes32 s) external {
+        rad.permit(owner, spender, value, deadline, v, r, s);
+
+        rad.burnFrom(msg.sender, fee);
+        commitments.commit(commitment);
+
+        emit CommitmentMade(commitment, block.number);
+
```
Comment (d-xo) [inline contracts/Registrar.sol]: nit: whitespace

File: contracts/Registrar.sol [hunk 2]
```diff
@@ -109,31 +102,44 @@ contract Registrar {
...
 
+    /// Commit to a future name and submit permit in the same transaction
+    function commitWithPermit(bytes32 commitment, address owner, address spender, uint256 value,
+                              uint256 deadline, uint8 v, bytes32 r, bytes32 s) external {
+        rad.permit(owner, spender, value, deadline, v, r, s);
+
+        rad.burnFrom(msg.sender, fee);
+        commitments.commit(commitment);
+
+        emit CommitmentMade(commitment, block.number);
```
Comment (d-xo) [inline contracts/Registrar.sol]: I think we can use the existing `commit` method here, which also checks to make sure that commitments are not being reused (I think there is a griefing vector where an attacker can continuously resubmit someone elses commitment, ensuring that they can never register their name).

File: contracts/Registrar.sol [hunk 3]
```diff
@@ -109,31 +102,44 @@ contract Registrar {
...
-        require(rad.balanceOf(msg.sender) >= fee, "Registrar::register: insufficient rad balance");
 
         rad.burnFrom(msg.sender, fee);
         commitments.commit(commitment);
 
         emit CommitmentMade(commitment, block.number);
     }
 
+    /// Commit to a future name and submit permit in the same transaction
+    function commitWithPermit(bytes32 commitment, address owner, address spender, uint256 value,
```
Comment (d-xo) [inline contracts/Registrar.sol:128]: Don't we also need to have a signature on the `commitment`?

File: contracts/Registrar.sol (hunk 2 above)
Comment (d-xo) [reply contracts/Registrar.sol]: although perhaps that attack doesn't work if we have a signature on the commitment.... :thinking:

Comment (MrChico) [reply contracts/Registrar.sol]: ah, yes, you are right that  `require(commitments.commited(commitment) == 0, "Registrar::commit: already commited"); ` is still missing. 
Notice that we still ask for payment from `msg.sender` here, so changing the commitment would do you no good

Comment (d-xo) [reply contracts/Registrar.sol]: > Notice that we still ask for payment from msg.sender here, so changing the commitment would do you no good

true!

File: contracts/Registrar.sol (hunk 3 above)
Comment (MrChico) [reply contracts/Registrar.sol:128]: we do for the 0 user actions pattern. For user initiated registrations `msg.sender` still pays, so it's fine

Comment (d-xo) [reply contracts/Registrar.sol:128]: or actually this is not required if we are burning from `msg.sender`

Comment (d-xo) [reply contracts/Registrar.sol:128]: hmmmm yes it's true. I think there is still a griefing vector in the 1 tx workflow without a signed commitment, although the attacker is going to end up having to register a lot of domains.... :thinking:

**Review by d-xo — APPROVED** [review]

**Pull request merged**